### DISCUSSION TITLE
How can i delete specific message of chat

### DISCUSSION THREAD

**Discussion by Xiddiqui** [tjtanjin/react-chatbotify#68] [General]
Hello .  I need help to delete specific message from chat . I tried customMessages as guided in documentation but it breaks my current implimentation
  const flow = {
        start: {
            message: "Hello! How can I assist you today?",
            options: ["Add Client KYC"],
            path: (input) => {
                if (input.userInput === "Add Client KYC") {
                    return "form_fill";
                } else {
                    return "loop";
                }
            },
        },
        loop: {
            message: async (params) => {
                const result = await fetchChatBotData(params.userInput);
                return result;
            },
            path: "loop",
        },
        form_fill: {
            message: "Please Fill the form bellow",
            render: (
                <ChatKycForm
                    loading={isLoading}
                    onSubmit={addLead}
                    onClose={() => setPaths((prev) => [...prev, "start"])}
                />
            ),
            chatDisabled: true,
        },
        end: {
            message: "Thank you! Your Lead created successfully",
            options: ["Add Client KYC"],
            path: (input) => {
                if (input.userInput === "Add Client KYC") {
                    return "form_fill";
                } else {
                    return "loop";
                }
            },
        },
        This is my flow and inside form_fill there is a component <ChatKycForm/> inside chatkycform there is button cancel when user press this button the last message from array should be deleted 
        How can i achive this

**Comment by tjtanjin** [comment]
Hey @Xiddiqui, could you show the code snippet for how you're using custom messages? Are you doing something similar to this:

```
setMessages(prevMessages => prevMessages.slice(0, -1));
```

Would also help if you can share more information on what you mean by break your current implementation.

**Reply by Xiddiqui** [reply]
Hey @tjtanjin . In my flow function . I have add when to render loop and when render form_fill  block so thats why i was unable to manage them when i created my array and push messages inside it 
const flow = {
start: {
message: "Hello! How can I assist you today?",
options: ["Add Client KYC"],
path: (input) => {
if (input.userInput === "Add Client KYC") {
return "form_fill";
} else {
return "loop";
}
},
},
loop: {
message: async (params) => {
const result = await fetchChatBotData(params.userInput);
return result;
},
path: "loop",
},
form_fill: {
message: "Please Fill the form bellow",
render: (
```
<ChatKycForm
loading={isLoading}
onSubmit={addLead}
onClose={() => setPaths((prev) => [...prev, "start"])}
/>
),
chatDisabled: true,
},
end: {
message: "Thank you! Your Lead created successfully",
options: ["Add Client KYC"],
path: (input) => {
if (input.userInput === "Add Client KYC") {
return "form_fill";
} else {
return "loop";
}
},
},
```

**Reply by tjtanjin** [reply]
Hey @Xiddiqui, components provided by `render` are also considered Messages. In fact every element that appears in the chatbot body are considered messages. You would be able to manage them all the same way. If it's not working for you, perhaps you can describe the following:
- What is the current behavior?
- What is the expected behavior?

If screenshots/videos will help to explain things feel free to include them as well.

**Reply by Xiddiqui** [reply]
Hi @tjtanjin 
<img width="406" alt="Screenshot 2024-05-30 at 10 00 26 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/5d9f381d-e0b5-44a0-9927-baf3ffc8d7cd">
In render i am showing this form and on cancel button i want to delete this message 
<img width="643" alt="Screenshot 2024-05-30 at 10 01 23 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/8edf5ea0-02f9-43f9-b0e9-1a7c061263df">
My Flow function looks like this 
<img width="441" alt="Screenshot 2024-05-30 at 10 01 44 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/4ae2846e-945a-4a65-a456-f84bb2ddbfcb">
Here i call chatBot and now i am confused how can i use both message context and path context. Is there any simplest way to achive this ?

**Reply by tjtanjin** [reply]
Hey @Xiddiqui, I would first wrap `ChatBot` with the message provider as well (currently only seeing path provider in your code). While doing that, you would have declared `messages` and `setMessages`, similar to what you did for paths. When the user cancels, you'll use `setMessages` to remove the last element in the array. Does this explanation make sense for you?

**Reply by Xiddiqui** [reply]
Okay Thanks for explainig me soo nicely @tjtanjin . I tried to implement as you explain but my issue is not fixed unfortunatly .
<img width="690" alt="Screenshot 2024-05-30 at 10 34 18 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/7c9ce37e-2368-4ddd-9308-8607023ed712">
Now I am Wraping chatbot with both contex
<img width="582" alt="Screenshot 2024-05-30 at 10 59 17 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/0ed99cbf-84c3-471c-a10d-7fe822aed4ff">

here i am calling insert message function 
<img width="386" alt="Screenshot 2024-05-30 at 10 34 46 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/793b37eb-aa88-4697-b550-1641ebfebe04">
my insert function is this

https://github.com/tjtanjin/react-chatbotify/assets/70687212/4846e98d-cb7c-4d08-8d9a-25cb69ccd191

Its not showing message that i added . Please guide me

**Reply by tjtanjin** [reply]
So a few things to note:
- The `path` attribute expects a string or a function that returns a string. The `insertMessage` you have above is a function that returns nothing which is why you observe your chatbot just stopping abruptly.
- I noticed that in this case, you're trying to add a message, not remove the last entry. If that is the case, you can use the `injectMessage` params shown here: https://react-chatbotify.tjtanjin.com/docs/api/params

The issue will be resolved if you address the 2 points above.

**Reply by Xiddiqui** [reply]
Okay okay . I was assuming that we need to add that message manualy in my array then delete it . But now i understand all message are in messages array that are managing through messageContext .So i remove code that i added for insert message and now here is what i am doing to delete 
<img width="466" alt="Screenshot 2024-05-30 at 11 14 09 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/20744c3c-402a-4b8b-a0db-8404bcc2965f">

https://github.com/tjtanjin/react-chatbotify/assets/70687212/e22e119a-96cc-45d9-a99b-365281d3c1ae

@tjtanjin  Appologies for disturbing you again and again

**Reply by tjtanjin** [reply]
Can I confirm you're still wrapping `ChatBot` with message provider? Could you also show what's inside `ChatKycForm`? As a sanity check you can try `setMessages([])` and see if messages are all cleared. If this doesn't work then it might be that `onClose` is not even being called.

**Reply by Xiddiqui** [reply]
@tjtanjin yes i wrapping it 
<img width="613" alt="Screenshot 2024-05-30 at 11 56 43 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/f5e21742-2ab6-443c-9780-3718c90d3ed1">
my KYC form is 
<img width="905" alt="Screenshot 2024-05-30 at 11 59 03 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/88893e4f-9744-41d1-ab09-fdb169dd8fb9">
here is my cancel button
<img width="965" alt="Screenshot 2024-05-30 at 11 59 14 AM" src="https://github.com/tjtanjin/react-chatbotify/assets/70687212/07a8f756-f9aa-44df-8203-6cb29ad49554">

This the start of my component

Additionaly i tried setMessages([]) , Its not clearing all messages

**Reply by tjtanjin** [reply]
What do you observe if you do `setMessages([])` instead?

**Reply by Xiddiqui** [reply]
I did it .My issue fixed now. Thank you @tjtanjin .
